[3006.x] fix(saltclass): don't lose nested classes and states in recursion

baby-gnu opened the pull request:
### What does this PR do?

This PR keep all the classes and states from recursive expansion of saltclass classes.

### What issues does this PR fix or reference?
Fixes #58969
Fixes #63933

### Previous Behavior

Only states from top level classes were kept.

### New Behavior

All states from all classes are used, in the correct order.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices, including the
[PR Guidelines](https://docs.saltproject.io/en/master/topics/development/pull_requests.html).

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

twangboy commented on tools/precommit/docstrings.py:
```diff
@@ -345,7 +345,6 @@
         "get_pillars",
         "expand_variables",
         "render_jinja",
-        "expand_classes_in_order",
```
Why is this removed?

baby-gnu replied:
Because I added a docsting to `expand_classes_in_order`.

When I looked up this function name in a global grep to see if any API need to be changed I found it there and I thought it was a good thing to do but I can remove that commit if it's inappropriate.

twangboy replied:
Oh, my bad... this is right

twangboy commented:
We'll need a changelog and some tests for this.

baby-gnu commented:
> We'll need a changelog and some tests for this.

Ok, the changelog is quite easy, I'll need to document for testing.

baby-gnu committed:
Add test for saltclass nested classes expansion

baby-gnu committed:
fix(saltclass): don't lose nested classes and states in recursion

Fixes: #58969

baby-gnu committed:
Fix incorrect override of minion node definition

A pillars dict in a minion node definition was not overridding
correctly the same dict in classes.

* salt/utils/saltclass.py (expanded_dict_from_minion): do not pass a
  reference to minion dict or it will be overridden by classes during
  expansion.

Fixes: #63933

baby-gnu committed:
Remove expand_classes_in_order from missing_docstrings

baby-gnu commented:
> We'll need a changelog and some tests for this.

@twangboy: I added tests for both issues:

<details><summary>Errors when passing the tests before my PR<code>nox -e 'test-3(coverage=False)' -- tests/pytests/unit/pillar/test_saltclass.py</code></summary>

```
nox > Running session test-3(coverage=False)
nox > Re-using existing virtual environment at .nox/test-3-coverage-false.
nox > Session test-3(coverage=False) was successful.
nox > Running session test-parametrized-3(crypto=None, transport='zeromq', coverage=False)
nox > Re-using existing virtual environment at .nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false.
nox > python -m pip install --progress-bar=off -U setuptools pip wheel
nox > python -m pip install --progress-bar=off -r requirements/static/ci/py3.10/linux.txt
nox > python -m pytest --rootdir /src --log-file-level=debug --show-capture=no -ra -s -vv --showlocals --log-file=/src/artifacts/logs/runtests-20240927064217.175879.log --transport=zeromq tests/pytests/unit/pillar/test_saltclass.py
<frozen importlib._bootstrap>:914: ImportWarning: TornadoImporter.find_spec() not found; falling back to find_module()
<frozen importlib._bootstrap>:914: ImportWarning: NaclImporter.find_spec() not found; falling back to find_module()
============================= test session starts ==============================
platform linux -- Python 3.10.14, pytest-8.1.1, pluggy-1.4.0 -- /src/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/bin/python
cachedir: .pytest_cache
max open files: soft=1048576; hard=1048576
salt-transport: zeromq
rootdir: /src
configfile: pytest.ini
plugins: skip-markers-1.5.2, httpserver-1.0.8, helpers-namespace-2021.4.29, flaky-3.8.1, salt-factories-1.0.1, timeout-2.3.1, pyfakefs-5.3.1, system-statistics-1.0.2, custom-exit-code-0.3.0, subtests-0.4.0, shell-utilities-1.8.0
collecting ... collected 7 items

tests/pytests/unit/pillar/test_saltclass.py::test_classes_order FAILED
tests/pytests/unit/pillar/test_saltclass.py::test_list_expansion_succeeds PASSED
tests/pytests/unit/pillar/test_saltclass.py::test_node_override_classes_scalars PASSED
tests/pytests/unit/pillar/test_saltclass.py::test_node_override_classes_scalar_in_dict FAILED
tests/pytests/unit/pillar/test_saltclass.py::test_node_override_classes_list_in_dict PASSED
tests/pytests/unit/pillar/test_saltclass.py::test_list_in_dict_no_duplication FAILED
tests/pytests/unit/pillar/test_saltclass.py::test_nested_classes_has_pillars PASSED

=================================== FAILURES ===================================
______________________________ test_classes_order ______________________________

temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_classes_order0/saltclass/examples')

    def test_classes_order(temp_saltclass_tree):
        """
        Classes must be correctly ordered.
    
        See: https://github.com/saltstack/salt/issues/58969
        """
        expected_ret = [
            "default.users.foo",
            "default.users",
            "default.motd",
            "default.empty",
            "default",
            "roles.app",
            "roles.nginx",
        ]
        fake_args = {"path": str(temp_saltclass_tree)}
        fake_pillar = {}
        fake_minion_id = "fake_id"
        try:
            full_ret = saltclass.ext_pillar(fake_minion_id, fake_pillar, fake_args)
            parsed_ret = full_ret["__saltclass__"]["classes"]
        # Fail the test if we hit our NoneType error
        except TypeError as err:
            pytest.fail(err)
        # Else give the parsed content result
>       assert expected_ret == parsed_ret
E       AssertionError: assert ['default.users.foo', 'default.users', 'default.motd', 'default.empty', 'default', 'roles.app', 'roles.nginx'] == ['default.users', 'default.motd', 'default.empty', 'default', 'roles.app', 'roles.nginx']
E         
E         At index 0 diff: 'default.users.foo' != 'default.users'
E         Left contains one more item: 'roles.nginx'
E         
E         Full diff:
E           [
E         +     'default.users.foo',
E               'default.users',
E               'default.motd',
E               'default.empty',
E               'default',
E               'roles.app',
E               'roles.nginx',
E           ]

expected_ret = ['default.users.foo',
 'default.users',
 'default.motd',
 'default.empty',
 'default',
 'roles.app',
 'roles.nginx']
fake_args  = {'path': '/tmp/pytest-of-root/pytest-77/test_classes_order0/saltclass/examples'}
fake_minion_id = 'fake_id'
fake_pillar = {'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from nginx',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
full_ret   = {'__saltclass__': {'classes': ['default.users',
                               'default.motd',
                               'default.empty',
                               'default',
                               'roles.app',
                               'roles.nginx'],
                   'environment': 'base',
                   'nodename': 'fake_id',
                   'states': ['users',
                              'motd',
                              'default',
                              'app',
                              'nginx',
                              'minion_node']},
 'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from minion_node',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
parsed_ret = ['default.users',
 'default.motd',
 'default.empty',
 'default',
 'roles.app',
 'roles.nginx']
temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_classes_order0/saltclass/examples')

tests/pytests/unit/pillar/test_saltclass.py:174: AssertionError
__________________ test_node_override_classes_scalar_in_dict ___________________

temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_node_override_classes_sca1/saltclass/examples')

    def test_node_override_classes_scalar_in_dict(temp_saltclass_tree):
        """
        Scalars defined in `dict` pillars defined in a node definition must override the
        same dict definition from classes.
    
        See: https://github.com/saltstack/salt/issues/63933
        """
        expected_ret = "from minion_node"
        fake_args = {"path": str(temp_saltclass_tree)}
        fake_pillar = {}
        fake_minion_id = "fake_id"
        try:
            full_ret = saltclass.ext_pillar(fake_minion_id, fake_pillar, fake_args)
            parsed_ret = full_ret["test_dict"]["a_scalar"]
        # Fail the test if we hit our NoneType error
        except TypeError as err:
            pytest.fail(err)
        # Else give the parsed content result
>       assert expected_ret == parsed_ret
E       AssertionError: assert 'from minion_node' == 'from users'
E         
E         - from users
E         + from minion_node

expected_ret = 'from minion_node'
fake_args  = {'path': '/tmp/pytest-of-root/pytest-77/test_node_override_classes_sca1/saltclass/examples'}
fake_minion_id = 'fake_id'
fake_pillar = {'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from nginx',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
full_ret   = {'__saltclass__': {'classes': ['default.users',
                               'default.motd',
                               'default.empty',
                               'default',
                               'roles.app',
                               'roles.nginx'],
                   'environment': 'base',
                   'nodename': 'fake_id',
                   'states': ['users',
                              'motd',
                              'default',
                              'app',
                              'nginx',
                              'minion_node']},
 'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from minion_node',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
parsed_ret = 'from users'
temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_node_override_classes_sca1/saltclass/examples')

tests/pytests/unit/pillar/test_saltclass.py:231: AssertionError
_______________________ test_list_in_dict_no_duplication _______________________

temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_list_in_dict_no_duplicati0/saltclass/examples')

    def test_list_in_dict_no_duplication(temp_saltclass_tree):
        """
        `list` under a `dict` in pillar must not be duplicated.
    
        See:
        """
        expected_ret = ["element1", "element2"]
        fake_args = {"path": str(temp_saltclass_tree)}
        fake_pillar = {}
        fake_minion_id = "fake_id"
        try:
            full_ret = saltclass.ext_pillar(fake_minion_id, fake_pillar, fake_args)
            parsed_ret = full_ret["test_dict"]["a_list"]
        # Fail the test if we hit our NoneType error
        except TypeError as err:
            pytest.fail(err)
        # Else give the parsed content result
>       assert expected_ret == parsed_ret
E       AssertionError: assert ['element1', 'element2'] == ['element1', 'element2', 'element1', 'element2']
E         
E         Right contains 2 more items, first extra item: 'element1'
E         
E         Full diff:
E           [
E               'element1',
E               'element2',
E         -     'element1',
E         -     'element2',
E           ]

expected_ret = ['element1', 'element2']
fake_args  = {'path': '/tmp/pytest-of-root/pytest-77/test_list_in_dict_no_duplicati0/saltclass/examples'}
fake_minion_id = 'fake_id'
fake_pillar = {'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from nginx',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
full_ret   = {'__saltclass__': {'classes': ['default.users',
                               'default.motd',
                               'default.empty',
                               'default',
                               'roles.app',
                               'roles.nginx'],
                   'environment': 'base',
                   'nodename': 'fake_id',
                   'states': ['users',
                              'motd',
                              'default',
                              'app',
                              'nginx',
                              'minion_node']},
 'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from minion_node',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
parsed_ret = ['element1', 'element2', 'element1', 'element2']
temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_list_in_dict_no_duplicati0/saltclass/examples')

tests/pytests/unit/pillar/test_saltclass.py:272: AssertionError
=========================== short test summary info ============================
FAILED tests/pytests/unit/pillar/test_saltclass.py::test_classes_order - Asse...
FAILED tests/pytests/unit/pillar/test_saltclass.py::test_node_override_classes_scalar_in_dict
FAILED tests/pytests/unit/pillar/test_saltclass.py::test_list_in_dict_no_duplication
========================= 3 failed, 4 passed in 1.09s ==========================
nox > Command python -m pytest --rootdir /src --log-file-level=debug --show-capture=no -ra -s -vv --showlocals --log-file=/src/artifacts/logs/runtests-20240927064217.175879.log --transport=zeromq tests/pytests/unit/pillar/test_saltclass.py failed with exit code 1
nox > Session test-parametrized-3(crypto=None, transport='zeromq', coverage=False) failed.
nox > Ran multiple sessions:
nox > * test-3(coverage=False): success
nox > * test-parametrized-3(crypto=None, transport='zeromq', coverage=False): failed
```
</details>

baby-gnu commented:
Thanks @twangboy, should I rebase the branch over 3006.x from time to time until it get merge?

baby-gnu commented:
> Thanks @twangboy, should I rebase the branch over 3006.x from time to time until it get merge?

I reword my commit message to fix typo as you did in the title :+1:
